### PR TITLE
Latent Dirichlet Allocation

### PR DESCRIPTION
**Pull request by Niklas-23**
Latent Dirichlet Allocation model implementation

Further features:
- Improved DataReader. The DataReader needs now only ~13.7 GB memory
- Pre-processor for tag removal using BeautifulSoup
- Per-processor for tokenization and stopword removal using nltk

### CODE REVIEW

File: src/pre_processors/nltk_tokenization_and_stopword_removal.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from src.base.query_pre_processor import QueryPreProcessor
+
+nltk.download('stopwords')
+nltk.download('punkt')
+stop_word_set = set(stopwords.words('english'))
+
+
+def tokenize_text(text: str) -> str:
+    text_tokens = word_tokenize(text)
+    tokens_without_sw = [word for word in text_tokens if not word in stop_word_set]
```
Comment (lkno0705) [inline src/pre_processors/nltk_tokenization_and_stopword_removal.py:19]: If you really want to remove all stop words, you have to make a lower case conversion to the target string, as the stop words in nltk are only in lowercase. So you'll miss Uppercase stop words here

File: src/pre_processors/nltk_tokenization_and_stopword_removal.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    filtered_sentence = (" ").join(tokens_without_sw)
+    return filtered_sentence
+
+
+class NLTKTokenizationAndStopwordRemoval(PreProcessor):
+
+    def forward(self, queries: List[Topic], documents: List[Union[Question, Answer]]) -> List[Union[Question, Answer]]:
+        for document in documents:
+            document.body = tokenize_text(document.body)
```
Comment (lkno0705) [inline src/pre_processors/nltk_tokenization_and_stopword_removal.py:28]: Question titles are left unprocessed

File: src/pre_processors/nltk_tokenization_and_stopword_removal.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        return documents
+
+
+class NLTKTokenizationAndStopwordRemovalForQueries(QueryPreProcessor):
+
+    def forward(self, queries: List[Topic]) -> List[Topic]:
+        for topic in queries:
+            topic.question = tokenize_text(topic.question)
```
Comment (lkno0705) [inline src/pre_processors/nltk_tokenization_and_stopword_removal.py:39]: Topic Bodys are left unprocessed

File: src/pre_processors/remove_xml_tags.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+from arqmath_code.topic_file_reader import Topic
+from src.base.pre_processing import PreProcessor
+from src.base.query_pre_processor import QueryPreProcessor
+
+
+def remove_xml_tags(text: str) -> str:
+    return BeautifulSoup(text, features="lxml").text
+
+
+class RemoveXMLTagsFromDocumentBody(PreProcessor):
```
Comment (lkno0705) [inline src/pre_processors/remove_xml_tags.py:15]: Same Problem as in the NLTK stopwords remover

**Review by lkno0705 — CHANGES_REQUESTED** [review]

File: src/pre_processors/nltk_tokenization_and_stopword_removal.py (hunk 1 above)
Comment (Niklas-23) [reply src/pre_processors/nltk_tokenization_and_stopword_removal.py:19]: fixed

**Review by lkno0705 — APPROVED** [review]
lgtm

**Pull request merged**